README.adoc

```diff
@@ -0,0 +1,411 @@
+
+**Fail**
+
+```js
+const sum = () => {
+  const numbers = Array.prototype.slice.call(arguments)
+  return numbers.reduce((a, b) => a + b)
+}
+
+sum(1, 2, 3)
+```
+
+**Pass**
+
+```js
+const sum (numbers) => (
+  numbers.reduce((a, b) => a + b)
+)
+
+sum([1, 2, 3])
+
+const args = node.arguments
+```
+
+## `fp/no-class`
+
+> Classes are nice tools to use when programming with the object-oriented paradigm, as they hold internal state and give access to methods on the instances. In functional programming, having stateful objects is more harmful than helpful, and should be replaced by the use of pure functions.
+
+Further reading: https://github.com/joshburgess/not-awesome-es6-classes/[Not Awesome: ES6 Classes; A curated list of resources on why ES6 (aka ES2015) classes are NOT awesome]
+
+**Fail**
+
+```js
+class Polygon {
+  constructor (height, width) {
+    this.height = height
+    this.width = width
+  }
+}
+```
+
+**Pass**
+
+```js
+const polygon = (height, width) => ({
+  height: height,
+  width: width,
+})
+```
+
+## `fp/no-delete`
+
+> delete is an operator to remove fields from an object or elements from an array. This purposely mutates data, which is not wanted when doing functional programming.
+
+Further reading: https://github.com/google/google-api-nodejs-client/issues/375[Avoid using delete operator]
+
+**Fail**
+```js
+delete foo
+delete foo.bar
+delete foo[bar]
+```
+**Pass**
+```js
+import F from "ramda"
+
+const fooWithoutBar = F.omit(["bar"], foo)
+const fooWithoutField = F.omit([bar], foo)
+```
+
+## `fp/no-events`
+
+> The use of EventEmitter with the events module provided by Node.js promotes implicit side-effects by emitting and listening to events. Instead of events, you should prefer activating the wanted effects by calling the functions you wish to use explicitly.
+
+Probably what you should do is use a https://gist.github.com/staltz/868e7e9bc2a7b8c1f754[functional reactive programming] library: https://github.com/cujojs/most[`most`], https://github.com/Reactive-Extensions/RxJS[`rxjs`].
+
+**Fail**
+
+```js
+import EventEmitter from "events"
+```
+
+## `fp/no-get-set`
+
+**Fail**
+```js
+const person = {
+  name: 'Some Name',
+  get age () {
+    return this._age
+  },
+  set age (n) {
+    if (n < 0) {
+      this._age = 0
+    } else if (n > 100) {
+      this._age = 100
+    } else {
+      this._age = n
+    }
+  }: 20
+};
+
+person.__defineGetter__("name", function () {
+  return this.name || "John Doe";
+})
+
+person.__defineSetter__("name", function (name) {
+  this.name = name.trim();
+})
+```
+**Pass**
+```js
+import F from "ramda"
+
+const person = {
+  name: "Some Name",
+  age: 20,
+}
+
+const clamp = (n, min, max) => (
+  n <= min ? min :
+  n >= max ? max :
+             n
+)
+
+const setAge = (age, person) => (
+  F.merge(person, { age: clamp(age, 0, 100) })
+)
+```
+
+## `fp/no-let`
```

Comment:
I don't fully agree. Yeah I'm aware of the mocha use-case. Probably like you I've also encountered use-cases outside mocha, in production code. But there are plenty of cases where `let` is bad design. It certainly complicates code reading. What to do? I think leverage the rule to help avert patterns that subvert the programmers best intentions. If the programmer knows better, they override the rule. Simple to do, high ROI.

In the case of tests, a module override could make sense:

```
/* eslint-disable fp/no-let */
```